lib/ansible/utils/plugin_docs.py

```diff
@@ -0,0 +1,161 @@
+
+from ansible.compat.six import string_types
+from ansible.parsing.yaml.loader import AnsibleLoader
+from ansible.plugins import fragment_loader
+
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+# modules that are ok that they do not have documentation strings
+BLACKLIST_MODULES = frozenset((
+    'async_wrapper',
+))
+
+
+def add_fragments(doc, filename):
+
+    fragments = doc.get('extends_documentation_fragment', [])
+
+    if isinstance(fragments, string_types):
+        fragments = [ fragments ]
+
+    # Allow the module to specify a var other than DOCUMENTATION
+    # to pull the fragment from, using dot notation as a separator
+    for fragment_slug in fragments:
+        fragment_slug = fragment_slug.lower()
+        if '.' in fragment_slug:
+            fragment_name, fragment_var = fragment_slug.split('.', 1)
+            fragment_var = fragment_var.upper()
+        else:
+            fragment_name, fragment_var = fragment_slug, 'DOCUMENTATION'
+
+        fragment_class = fragment_loader.get(fragment_name)
+        assert fragment_class is not None
+
+        fragment_yaml = getattr(fragment_class, fragment_var, '{}')
+        fragment = AnsibleLoader(fragment_yaml, file_name=filename).get_single_data()
+
+        if 'notes' in fragment:
+            notes = fragment.pop('notes')
+            if notes:
+                if 'notes' not in doc:
+                    doc['notes'] = []
+                doc['notes'].extend(notes)
+
+        if 'options' not in fragment:
+            raise Exception("missing options in fragment (%s), possibly misformatted?: %s" % (fragment_name, filename))
+
+        for key, value in fragment.items():
+            if key not in doc:
+                doc[key] = value
+            else:
+                if isinstance(doc[key], MutableMapping):
+                    doc[key].update(value)
+                elif isinstance(doc[key], MutableSet):
+                    doc[key].add(value)
+                elif isinstance(doc[key], MutableSequence):
+                    doc[key] = sorted(frozenset(doc[key] + value))
+                else:
+                    raise Exception("Attempt to extend a documentation fragement (%s) of unknown type: %s" % (fragment_name, filename))
+
+
+def get_docstring(filename, verbose=False):
+    """
+    Search for assignment of the DOCUMENTATION and EXAMPLES variables
+    in the given file.
+    Parse DOCUMENTATION from YAML and return the YAML doc or None
+    together with EXAMPLES, as plain text.
+
+    DOCUMENTATION can be extended using documentation fragments
+    loaded by the PluginLoader from the module_docs_fragments
+    directory.
+    """
+
+    data = {
+        'doc': None,
+        'plainexamples': None,
+        'returndocs': None,
+        'metadata': {'metadata_version': '1.0', 'status': ['preview'], 'supported_by': 'community'},
```

Comment:
Do we need default metadata? I wonder if in the future we can use CI to ensure this is present in all files, not just modules.